### ISSUE TITLE
Non-ASCII characters break channel switching

### ISSUE THREAD

**Issue by BaerMitUmlaut** [IDI-Systems/acre2#579]
**Arma 3 Version:** `1.84` stable
**CBA Version:** `3.8.0` stable
**ACRE2 Version:** `2.6.1` stable

**Mods:**
```
- CBA_A3
- acre2
```

**Description:**
Non-ASCII characters in the server name breaks the channel switching system.

**Steps to reproduce:**
- [Sample server.cfg](https://files.slack.com/files-pri/T0J06UTRP-FCNEBG5A7/download/server.cfg)
- `hostName = "7.Kolordu Komutanlığı Teamspeak (ACRE) | 7KolorduTS | ";`

**Where did the issue occur?**
- Dedicated

**Placed Modules:**
- None

Issue created for max315 from Slack.

**Comment by jonpas** [comment]
@tbeswick96 ?

**Comment by tbeswick96** [comment]
@jonpas not sure I can find the time to get round to this....I can barely find the time to play with my own community at the moment :/

**Comment by jonpas** [comment]
@tbeswick96 ok, no problem, anyone else wants to pitch in, go ahead! :D

**Comment by bovine3dom** [comment]
Briefly looked into this. UTF-8 strings in C++ should "just work" AFAICT. Documenting my findings here:

`requestClientMove` is the TS3 function that moves you round channels.

`ACRE_RESULT CTS3Client::moveToServerTS3Channel()` moves you to the "ACRE activated" / specified channel.

`getTs3ChannelDetails` looks like it tells you the name of the specified channel. It's cunningly hidden in a macro:

```
extensions\src\ACRE2TS\TS3Client.h
97:    DECLARE_MEMBER(std::vector<std::string>, Ts3ChannelDetails);
```

Which leads you to yet more macros that just give you getters and setters that are 100% uninteresting:
```
#define DECLARE_MEMBER(type, name)                     \
public:                                                \
    DECLARE_MEMBER_SET(type, name)                    \
    DECLARE_MEMBER_GET(type, name)                    \
protected:                                            \
    DECLARE_MEMBER_PRIVATE(type, name)        
```

This means that `setTs3ChannelDetails` is the function we're looking for.

If C++ UTF-8 gives us too much trouble, https://github.com/nemtrif/utfcpp looks like a neat little library for doing it.

I'm hoping that the problem lies in the SQF side of things but I'm preparing for the worst.

To be continued...

**Comment by bovine3dom** [comment]
`updateTs3ChannelDetails(details)` in src/ACRE2Core/setTs3ChannelDetails.h seems to be where the channel name is got from ArmA.

On the SQF side, we have:

```
addons\sys_io\fnc_ts3ChannelMove.sqf
18:private _ts3ChannelDetails = format ["%1,%2,%3", EGVAR(sys_core,ts3ChannelName), EGVAR(sys_core,ts3ChannelPassword), serverName];
19:TRACE_1("Moving TS3 Channel",_ts3ChannelDetails);
20:CALL_RPC("setTs3ChannelDetails",_ts3ChannelDetails);
```

So, I think the mangling happens in one of three places:

1. In the initial formatting:
```
addons\sys_io\fnc_ts3ChannelMove.sqf
18:private _ts3ChannelDetails = format ["%1,%2,%3", EGVAR(sys_core,ts3ChannelName), EGVAR(sys_core,ts3ChannelPassword), serverName];
```

2. In the call to TS
```
addons\sys_io\fnc_ts3ChannelMove.sqf
20:CALL_RPC("setTs3ChannelDetails",_ts3ChannelDetails);
```

3. In the conversion of the `_ts3ChannelDetails` into a vector
```
extensions\src\ACRE2Core\setTs3ChannelDetails.h
10:RPC_FUNCTION(setTs3ChannelDetails) {
// mangling might happen here
    std::vector<std::string> details = {
        std::string((char *)vMessage->getParameter(0)),
        std::string((char *)vMessage->getParameter(1)),
        std::string((char *)vMessage->getParameter(2))
    };
///
17:    CEngine::getInstance()->getClient()->updateTs3ChannelDetails(details);
```

Logging `_ts3ChannelDetails` in `addons\sys_io\fnc_ts3ChannelMove.sqf` seems like the first (and easiest) step to working out what is being mangled and where.

I probably won't have time for a while, so anyone who wants to pick this up should do :)

**Comment by jonpas** [comment]
Fixed by #1326.
